### ISSUE TITLE
Is it possible to get scores from a search

### ISSUE THREAD

**Issue by rmcolq** [seqan/raptor#439]
This is likely an enhancement/feature request rather than already implemented.

When performing a raptor search with a read file against an index with e.g. 2 bins, is there any way of getting the number of kmers/minimizers which were found in each bin instead of a list of bins which meet threshold criteria? eg. read1 had 100 minimizers, of which 40 were found in the index for bin1 and 20 were found in the index for bin2? It seems like this information must be implicitly calculated and I have an application where it would be very useful. Even if this information is only calculated for those that meet the threshold criteria?

**Comment by eseiler** [comment]
Hey there,

we indeed have the counts and then do something with them.

#### IBF
For the IBF (just running `raptor build`), we get the counts and then filter them: 
<details><summary>Code Example (Click to expand)</summary>

https://github.com/seqan/raptor/blob/3e8eebc74f940802c16ff35332cf65d3bb8f2bc0/include/raptor/search/search_singular_ibf.hpp#L87-L104
</details> 

#### HIBF
For the HIBF (running `raptor layout` first, and then using the result as input for `raptor build`), there is an API that does this internally. The Hierarchical IBF (HIBF) uses the threshold internally to decide which parts of the hierarchy need to be traversed.
<details><summary>Code Example (Click to expand)</summary>

https://github.com/seqan/raptor/blob/3e8eebc74f940802c16ff35332cf65d3bb8f2bc0/include/raptor/search/search_singular_ibf.hpp#L104-L115
</details> 

There is also a count interface for the HIBF.

---

So, the counts do exist at some point, and we (as well as other people) have different applications for different tasks:
* [Metagenomic Classification](https://github.com/pirovc/ganon)
* [Read Mapping](https://github.com/seqan/dream_yara)
* [Quantification Estimation](https://github.com/seqan/needle)
* [Local Sequence Similarity Search](https://github.com/eaasna/valik)
* Some others that are under development

The (H)IBF is often used as a pre-filter (basically what Raptor does).

### Outputting Counts

It wouldn't be difficult to also add an option to output the counts. However, it would increase the amount of maintenance needed (unit tests, etc.). In my opinion, outputting counts tends to be rather not in the scope of Raptor.

Another considerable drawback is that outputting counts easily results in massive output files (In the range of multiple GiB or even a few TiB), because (at least choosing the easiest route) we would have to write a matrix:

```
#READ_ID BIN_1_COUNT BIN_2_COUNT ... BIN_X_COUNT
read_1   21          0               50
...
read_Y   5           34              12
```

Currently, I'm rather against adding this as a feature to raptor.

### What next?

It depends on your background. If you are familiar with C++, it wouldn't be difficult to write a small application that outputs the counts, or fork raptor and make the changes there. 

I also wouldn't mind preparing a version that just outputs the counts instead of the bins (without having these changes in the main branch).

Regarding the output files, I guess the size doesn't really matter for a proof of concept and the format could later be optimized.
If the counts are used to do something, it would be best to just do it in the application directly, if possible, and omit writing all the counts to disk.

So, in summary, here are the things that I could offer:

1. Creating a small application that outputs the counts (we would need to shortly discuss requirements).
2. Same as above, but with the goal of it being used for further development by you.
3. You fork Raptor, which is similar to 2. - but with much more code :)
4. I create a branch somewhere that replaces bin output with count output.
5. If you are not familiar with C++ and/or compiling from source, I could also provide a binary that incorporates the changes. I would probably add a GitHub action that uploads the binary as an artifact, such that it's transparent how the binary was generated (in contrast to just downloading some binary from a basically unknown source).
6. Any ideas you have.


Let me know what works for you.

**Comment by rmcolq** [comment]
Hi @eseiler thanks for the in depth reply - super helpful! I do have C++ experience but it was a long while ago - I have been trying to explore the code in seqan3 and raptor this week to see if I could fork/develop/adapt something that is similar but specific to my application.

I completely understand the overhead of having this (large alternative) output is out of scope for raptor. For my application I would probably want to be working with only 2 (or a small number of) bins and the ultimate goal would be to get to a probability score of a read belonging to group A taxa vs group B taxa (different taxonomic domains). But I need this to be fast (and raptor does seems to get most of the way fast). I don't mind so much about the size of the file output.

Thank you for the offers of development help! I think I need to explore the code for a bit longer and perhaps come back to you if I'm still stuck! 

I'll close this issue so it isn't clogging up your undone list and reopen as/when needed?
